cql3/restrictions/statement_restrictions.cc

```diff
@@ -1043,6 +1047,381 @@ expression make_conjunction(const expression& a, const expression& b) {
     return conjunction{children};
 }
 
+namespace {
+
+using cql3::selection::selection;
+
+/// Serialized values for all types of cells, plus selection (to find a column's index) and options (for
+/// subscript term's value).
+struct row_data {
+    const std::vector<bytes>& partition_key;
+    const std::vector<bytes>& clustering_key;
+    const std::vector<bytes_opt>& other_columns;
+    const selection& sel;
+    const query_options& options;
+};
+
+/// Returns col's value from the fetched data.
+bytes_opt get_value(const column_value& col, row_data data) {
+    auto cdef = col.col;
+    if (col.sub) {
+        auto col_type = static_pointer_cast<const collection_type_impl>(cdef->type);
+        if (!col_type->is_map()) {
+            throw exceptions::invalid_request_exception("subscripting non-map column");
+        }
+        const auto deserialized = cdef->type->deserialize(*data.other_columns[data.sel.index_of(*cdef)]);
```

Comment:
This index_of lookup is a strong candidate for future storage in the expression (or a transformation thereof).

Reply:
Agreed.